### ISSUE TITLE
[Forwarder] Private link support (VPC configuration support through CF template) 

### ISSUE THREAD

**Issue by rromanchuk** [DataDog/datadog-serverless-functions#225]
https://docs.datadoghq.com/agent/guide/private-link/?tab=logs#client-configuration

The above documentation is a little misleading since the default cloudformation template doesn't seem to support enabling VPC/subnets/security group. It also can't be added by brute force unless you attach the missing `AWSLambdaVPCAccessExecutionRole` to the role. 

I could be missing something, but it's not entirely obvious. Are you guys planning on adding these config parameters at some point? I can send send a PR, but this will introduce some pretty obnoxious branching/parameter collecting that you guys are going to able pump out a lot faster/cleaner.

`AWS::Serverless::Function` would need a branching  for`VpcConfig`
https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-lambda-function-vpcconfig.html 

Also the execution role will need aws managed `AWSLambdaVPCAccessExecutionRole`

I got it working, but i'm just drifting the stack and brute forcing my way there, just not a long term solution.

**Comment by tianchu** [comment]
@rromanchuk do you mind sharing us with some extra context and requirements for your use case ? E.g., what forces you to run Datadog forwarder Lambda inside an VPC? What parameters you wish to have to support your use case? We are happy to build that into a future version of the template.

**Comment by rromanchuk** [comment]
@tianchu  Yeah, my current hosts on ec2, using the log collector with the datadog-agent are now forwarding logs via `pvtlink.logs.datadoghq.com`,  which allows me to avoid sending outbound traffic through the www via my nat gateway, Now i use datadog's private link with my VPC in us-east-1.

I'd like to do the same for lambda forwarder. I'm not forced to, it just (seems to) make a lot of sense.  I've got it working fine, just a bit clunky since i'm drifting my stack. 

In general, data transfer is cheaper when it doesn't leave the aws ecosystem. This is especially true if you're routing outbound traffic through a NAT. If you have a webserver transfering data to s3 from within a private subnet, a region away from from the NAT gateway, you can save a lot on data transfer by avoiding those hops entirely. I haven't looked at the economics in context of private link/lambda data transfer, but it was more about being consistent and conscious about where traffic is flowing.

**Comment by rromanchuk** [comment]
I guess if you're being charged to provision the private link, and in the context of this lambda function,  maybe the economics end up being cheaper for both parties sending over an internet gateway, free inbound for datadog. There are also no dependent VPC resources required to run this function, where it would be necessary in the context of a traditional ec2 host logging. I guess i can see where this wouldn't be ideal for either party. Sorry just thinking about this on the fly, covid19 infrastructure spring cleaning...

**Comment by tianchu** [comment]
Thanks @rromanchuk, this makes perfect sense now. I have logged a ticket for me team to consider supporting PrivateLink OOB.

**Comment by pinkerton** [comment]
Hi @rromanchuk, I'm the product manager for the Serverless team here at Datadog. @tianchu and I spoke about this request, and we plan to prioritize it in an upcoming engineering sprint. One of us will follow up here when we have more details.

**Comment by dezren39** [comment]
This is exciting! I'm beginning on the exact same path as @rromanchuk over the past week, when I realized there was no VPC so it couldn't use my privatelink is when I searched and found this. I think I'll be manually modifying my lambdas in the mean-time, but definitely interested in seeing this update.

It's probably much more important for the forwarder as logs can end up being quite a few GB in some cases, but I'm also looking at it for the RDS enhanced metrics lambda even though it's just api/metrics, small traffic. Since all the agents and everything else would be using private link, it seems funny to just have the one going over the internet gateway.
